rholang/src/main/scala/coop/rchain/rholang/interpreter/error/RWLock.scala

```diff
@@ -0,0 +1,163 @@
+package coop.rchain.rholang.interpreter.error
+
+import cats.effect.concurrent.MVar
+import cats.effect.{Bracket, Concurrent}
+import cats.implicits._
+
+/**
+  * [[Free]]: Read or write lock can be acquired without blocking.
+  *
+  * [[Write]]: Read and write locks block until lock is released.
+  *
+  * [[Read]]: Write locks block until all read locks are released.
+  *
+  */
+sealed trait State
+case object Write                 extends State
+case object Free                  extends State
+final case class Read(count: Int) extends State
+
+final class RWLock[F[_]](stateRef: MVar[F, State], readLock: MLock[F], writeLock: MLock[F])(
+    implicit bracketF: Bracket[F, Throwable]
+) {
+
+  /**
+    * Acquires read lock.
+    *
+    * Blocks (asynchronously) if another thread has write lock.
+    *
+    */
+  def acquireRead: F[Unit] =
+    stateRef.take >>= {
+      case Free        => readLock.acquire >> stateRef.put(Read(1))
+      case Write       => stateRef.put(Write) >> writeLock.waitForLock >> acquireRead
```

Comment:
@lukasz-golebiewski This is why the wait functionality is required. This isn't infinite recursion as you suggested. In the case that the state reference is in the `Write` state, the `waitForLock` function asynchronously blocks the `acquireRead` function until the write lock is released. The only time one call to `acquireRead` recurses more than once is if another write permit is granted before it calls `stateRef.take`.

Reply:
The wait functionality is required because we want all readers to wake up simultaneously when the write lock is returned.

Reply:
I never suggested this *was* infinite recursion, I merely stated that this is what I see in the snippet you provided in https://github.com/rchain/rchain/pull/2504#discussion_r292542311 mainly due to the lack of context

Reply:
This PR and the tests make things clearer. Thanks!